### PR TITLE
Added free-carrier absorption (FCA) and free-carrier plasma dispersion (FCPD)

### PR DESCRIPTION
**Pull request by caseyflex**
Adds FCA and FCPD inside TwoPhotonAbsorption class.

### CODE REVIEW

File: tidy3d/components/medium.py [hunk 1]
```diff
@@ -229,13 +263,20 @@ def complex_fields(self) -> bool:
 class TwoPhotonAbsorption(NonlinearModel):
     """Model for two-photon absorption (TPA) nonlinearity which gives an intensity-dependent
     absorption of the form :math:`\\alpha = \\alpha_0 + \\beta I`.
+    Also includes free-carrier absorption (FCA) and free-carrier plasma (FCP) effects.
```
Comment (momchil-flex) [inline tidy3d/components/medium.py:266]: Is FCP the same as free-carrier dispersion (i.e. modulation of the real part of the permittivity)? I've heard the latter but not the former.

File: tests/test_components/test_medium.py [hunk 2]
```diff
@@ -624,7 +623,7 @@ def test_nonlinear_medium(log_capture):
     # active materials
     with pytest.raises(ValidationError):
         med = td.Medium(
-            nonlinear_spec=td.NonlinearSpec(models=[td.TwoPhotonAbsorption(beta=-1 + 1j, n0=1)])
+            nonlinear_spec=td.NonlinearSpec(models=[td.TwoPhotonAbsorption(beta=-1, n0=1)])
```
Comment (momchil-flex) [inline tests/test_components/test_medium.py:626]: What exactly is happening with `beta`? It used to be possible to set it as complex, where the imaginary part means modulation of the *real* part of the permittivity? But TPA is purely absorption, so now instead we want beta to be purely real, while any dispersion would be coming from FCP?

Is the medium still unstable if the nonlinear effect is purely lossy like TPA?

**Review by momchil-flex — COMMENTED** [review]
Some initial comments

File: tidy3d/components/medium.py (hunk 1 above)
Comment (caseyflex) [reply tidy3d/components/medium.py:266]: yeah, if that is more familiar I can change the terminology

File: tests/test_components/test_medium.py (hunk 2 above)
Comment (caseyflex) [reply tests/test_components/test_medium.py:626]: `beta` used to be complex, although typically a real beta would be used. I think you would use KerrNonlinearity instead if you wanted to modify the real part of the permittivity directly, or as you said you can have FCP dispersion coming from the same origin as the TPA.

Complex `beta` is implemented according to the docstring equation for TPA nonlinear polarization, which involves both the complex `n0` and `Re(n0)`. This equation is derived assuming real `beta`. For complex `beta`, we can use the same equation / implementation, but the physics may be less clear, and I cannot think of a use case. So, I deprecated complex `beta`.

TPA is lossy when `beta` is positive -- a negative value of `beta` would instead add energy in a nonlinear fashion, which may actually be used in some cases, although is an active medium and can lead to instability. The more general condition for stability involves the actual nonlinear polarization (and is what we use in the validator), and reduces to this when `n0` is real.

File: tests/sims/simulation_2_6_0rc1.json [hunk 3]
```diff
@@ -816,11 +816,23 @@
...
+                            "beta": 1.0,
+                            "tau": 1.0,
+                            "sigma": 1.0,
+                            "e_e": 1.0,
+                            "e_h": 0.8,
+                            "c_e": 1.0,
+                            "c_h": 1.0,
+                            "n0": null,
+                            "freq0": null
+>>>>>>> 26576d4d (Added free-carrier absorption (FCA) and free-carrier plasma dispersion (FCPD) nonlinearities.)
```
Comment (momchil-flex) [inline tests/sims/simulation_2_6_0rc1.json:835]: Not sure why there's a conflict here. Could you rerun test_local maybe and see if that would fix it (it generates this file). Also, `git checkout develop -- tests/sims/simulation_2_5_0.json`.

**Review by momchil-flex — APPROVED** [review]
This looks good, I'll look at the backend changes again too. Is this waiting for the notebook example to be done / work correctly though?

File: tests/sims/simulation_2_6_0rc1.json (hunk 3 above)
Comment (caseyflex) [reply tests/sims/simulation_2_6_0rc1.json]: I don't think it needs to wait for the notebook example...

The two simulation json files are fixed now.

**Pull request merged**